File: Back-End/backend/src/models/Transaction.ts
import { db } from '../global/firebase';
import { Timestamp, FieldValue } from 'firebase-admin/firestore';

export interface TransactionInput {
  stockId: string;
  type: 'buy' | 'sell';
  date: string; // ISO string
  shares: number;
  pricePerShare: number;
  fee: number;
  note?: string;
}

export class Transaction {
  id!: string;
  stockId!: string;
  type!: 'buy' | 'sell';
  date!: string;
  shares!: number;
  pricePerShare!: number;
  fee!: number;
  note!: string;
  createdAt!: string;

  private static readonly col = db.collection('transactions');

  static async findAll(stockId?: string): Promise<Transaction[]> {
    const baseQuery = stockId
      ? this.col.where('stock_id', '==', stockId)
      : this.col;
    const snap = await baseQuery.get();
    const items = snap.docs.map(doc => Transaction.fromSnapshot(doc));
    return items.sort((a, b) => a.date.localeCompare(b.date));
  }

  static async findById(id: string): Promise<Transaction | null> {
    const doc = await this.col.doc(id).get();
    return doc.exists ? Transaction.fromSnapshot(doc) : null;
  }

  static async create(input: TransactionInput): Promise<Transaction> {
    const docRef = await this.col.add({
      stock_id:        input.stockId,
      type:            input.type,
      date:            Timestamp.fromDate(new Date(input.date)),
      shares:          input.shares,
      price_per_share: input.pricePerShare,
      fee:             input.fee,
      note:            input.note ?? '',
      created_at:      FieldValue.serverTimestamp(),
    });
    const doc = await docRef.get();
    return Transaction.fromSnapshot(doc);
  }

  static async update(id: string, input: Partial<TransactionInput>): Promise<Transaction | null> {
    const ref = this.col.doc(id);
    const doc = await ref.get();
    if (!doc.exists) return null;

    const patch: Record<string, unknown> = {};
    if (input.stockId        !== undefined) patch['stock_id']        = input.stockId;
    if (input.type           !== undefined) patch['type']            = input.type;
    if (input.date           !== undefined) patch['date']            = Timestamp.fromDate(new Date(input.date));
    if (input.shares         !== undefined) patch['shares']          = input.shares;
    if (input.pricePerShare  !== undefined) patch['price_per_share'] = input.pricePerShare;
    if (input.fee            !== undefined) patch['fee']             = input.fee;
    if (input.note           !== undefined) patch['note']            = input.note;

    await ref.update(patch);
    const updated = await ref.get();
    return Transaction.fromSnapshot(updated);
  }

  static async delete(id: string): Promise<boolean> {
    const doc = await this.col.doc(id).get();
    if (!doc.exists) return false;
    await this.col.doc(id).delete();
    return true;
  }

  private static fromSnapshot(doc: FirebaseFirestore.DocumentSnapshot): Transaction {
    const d = doc.data()!;
    const t = new Transaction();
    t.id            = doc.id;
    t.stockId       = d['stock_id'];
    t.type          = d['type'];
    t.date          = (d['date'] as Timestamp).toDate().toISOString();
    t.shares        = d['shares'];
    t.pricePerShare = d['price_per_share'];
    t.fee           = d['fee'];
    t.note          = d['note'] ?? '';
    const ca        = d['created_at'];
    t.createdAt     = ca instanceof Timestamp ? ca.toDate().toISOString() : new Date().toISOString();
    return t;
  }
}

File: Back-End/shioaji-service/models/transaction.py
from __future__ import annotations
import asyncio
from typing import Optional
from google.cloud.firestore import SERVER_TIMESTAMP
from lib.firebase import get_db

COL = "transactions"


def _ts_to_iso(ts) -> str:
    if hasattr(ts, "isoformat"):
        return ts.isoformat()
    from datetime import datetime, timezone
    return datetime.now(tz=timezone.utc).isoformat()


def _from_doc(doc) -> dict:
    d = doc.to_dict()
    return {
        "id": doc.id,
        "stockId": d.get("stock_id", ""),
        "type": d.get("type", "buy"),
        "date": _ts_to_iso(d.get("date")),
        "shares": d.get("shares", 0),
        "pricePerShare": d.get("price_per_share", 0),
        "fee": d.get("fee", 0),
        "note": d.get("note", ""),
        "createdAt": _ts_to_iso(d.get("created_at")),
    }


async def find_all(stock_id: Optional[str] = None) -> list[dict]:
    def _run():
        col = get_db().collection(COL)
        if stock_id:
            query = col.where("stock_id", "==", stock_id)
        else:
            query = col
        snap = query.get()
        docs = [_from_doc(d) for d in snap]
        return sorted(docs, key=lambda t: t["date"])
    return await asyncio.to_thread(_run)


async def find_by_id(doc_id: str) -> Optional[dict]:
    def _run():
        doc = get_db().collection(COL).document(doc_id).get()
        return _from_doc(doc) if doc.exists else None
    return await asyncio.to_thread(_run)


async def create(data: dict) -> dict:
    def _run():
        db = get_db()
        ref = db.collection(COL).document()
        ref.set({
            "stock_id": data["stockId"],
            "type": data["type"],
            "date": _parse_ts(data["date"]) if data.get("date") else SERVER_TIMESTAMP,
            "shares": data["shares"],
            "price_per_share": data["pricePerShare"],
            "fee": data["fee"],
            "note": data.get("note", ""),
            "created_at": SERVER_TIMESTAMP,
        })
        return _from_doc(ref.get())
    return await asyncio.to_thread(_run)


async def update(doc_id: str, patch: dict) -> Optional[dict]:
    def _run():
        ref = get_db().collection(COL).document(doc_id)
        doc = ref.get()
        if not doc.exists:
            return None
        data: dict = {}
        if "stockId" in patch:       data["stock_id"]        = patch["stockId"]
        if "type" in patch:          data["type"]            = patch["type"]
        if "date" in patch:          data["date"]            = _parse_ts(patch["date"])
        if "shares" in patch:        data["shares"]          = patch["shares"]
        if "pricePerShare" in patch: data["price_per_share"] = patch["pricePerShare"]
        if "fee" in patch:           data["fee"]             = patch["fee"]
        if "note" in patch:          data["note"]            = patch["note"]
        ref.update(data)
        return _from_doc(ref.get())
    return await asyncio.to_thread(_run)


async def delete(doc_id: str) -> bool:
    def _run():
        ref = get_db().collection(COL).document(doc_id)
        if not ref.get().exists:
            return False
        ref.delete()
        return True
    return await asyncio.to_thread(_run)


def _parse_ts(iso_str: str):
    from datetime import datetime, timezone
    dt = datetime.fromisoformat(iso_str.replace("Z", "+00:00"))
    return dt
